File: 18-4sum/4sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
    let result = [];
    nums.sort((a, b) => (a-b))
    for(let i = 0; i<nums.length; i++){
        if(i > 0 && nums[i] == nums[i - 1]) continue;
        for(let j = i + 1; j<nums.length; j++){
            if(j > i + 1 && nums[j] == nums[j - 1]) continue;
            let l = j + 1;
            let r = nums.length - 1;
            while(l < r){
                let sum = nums[i] + nums[j] + nums[l] + nums[r];
                if(sum < target){
                    l++;
                }
                else if(sum > target){
                    r--;
                }
                else{
                    result.push([nums[i], nums[j], nums[l], nums[r]]);
                    l++;
                    r--;
                    while (l < r && nums[l] === nums[l - 1]) l++;
                    while (l < r && nums[r] === nums[r + 1]) r--;
                }
            }
        }
    }
    return result;
};

File: 18-4sum/4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        result = []
        
        for i in range(n-3):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            for j in range(i+1, n-2):
                if j > i + 1 and nums[j] == nums[j -1]:
                    continue
                
                s = j+1
                w = n-1
                while(s < w):
                    total = nums[i] + nums[j] + nums[s] + nums[w]

                    if total == target:
                        result.append([nums[i], nums[j], nums[s], nums[w]])
                        s+=1
                        w-=1
                        #skip duplicate of s 
                        while s < w and nums[s] == nums[s-1]:
                            s += 1
                        
                        while s < w and nums[w] == nums[w+1]:
                            w -= 1
                        

                    elif total > target:
                        w -= 1
                    else:
                        s += 1
        return result